### PR TITLE
Full feature integration test

### PR DESCRIPTION
**Pull request by agudian**
... Just a teaser.

As @sjaakd mentioned in PR #692, the following fails:

On JDK 6:
```
[INFO] Compiling 660 source files to C:\data\workspace-stuff4\mapstruct\integrationtest\target\tmp\org.mapstruct.itest.tests.FullFeatureCompilationTest\oracle_java_6\target\classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] ...\mapstruct\integrationtest\target\tmp\org.mapstruct.itest.tests.FullFeatureCompilationTest\oracle_java_6\target\generated-sources\annotations\org\mapstruct\ap\test\abstractclass\generics\GenericsHierarchyMapperImpl.java:[33,44] reference to getItem is ambiguous, both method getItem() in org.mapstruct.ap.test.abstractclass.generics.ItemProviderSomeItemA<org.mapstruct.ap.test.abstractclass.generics.ItemA> and method getItem() in org.mapstruct.ap.test.abstractclass.generics.ItemProviderAny<org.mapstruct.ap.test.abstractclass.generics.ItemB> match
[ERROR] ...\mapstruct\integrationtest\target\tmp\org.mapstruct.itest.tests.FullFeatureCompilationTest\oracle_java_6\target\generated-sources\annotations\org\mapstruct\ap\test\abstractclass\generics\GenericsHierarchyMapperImpl.java:[33,25] modifyItemB(org.mapstruct.ap.test.abstractclass.generics.ItemB) in org.mapstruct.ap.test.abstractclass.generics.GenericsHierarchyMapper cannot be applied to (org.mapstruct.ap.test.abstractclass.generics.ItemA)
[INFO] 2 errors 
```

And on JDK 9:
```
[INFO] Compiling 660 source files to C:\data\workspace-stuff4\mapstruct\integrationtest\target\tmp\org.mapstruct.itest.tests.FullFeatureCompilationTest\oracle_java_9\target\classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] ...\mapstruct\processor\src\test\java\org\mapstruct\ap\test\collection\wildcard\BeanMapper.java:[36,32] error: Can't map property "javax.xml.bind.JAXBElement<? extends java.math.BigDecimal> content" to "java.math.BigDecimal content". Consider to declare/implement a mapping method: "java.math.BigDecimal map(javax.xml.bind.JAXBElement<? extends java.math.BigDecimal> value)".
[INFO] 1 error
```

The test uses JDK 6, 7, 8, 9 (I have the latest preview installed), Eclipse JDT with language levels 1.6, 1.7, 1.8.

### CODE REVIEW

**Comment by sjaakd** [comment]
Hmm. Java 9 is a bit unexpected.. isn't it? But I reckon we have some time to fix that one before Java 9 is GA ( 22 september 2016) :smile:

**Comment by agudian** [comment]
Right, I think so, too... :grin:

**Comment by gunnarmorling** [comment]
@agudian Could you rebase this to the latest master?

**Comment by agudian** [comment]
@gunnarmorling, it was based on the commits of PR #692 - as far as I can see, it's now just the one commit ahead. :)

**Comment by agudian** [comment]
Commit rebased to current master.

**Comment by agudian** [comment]
retest this, please

File: integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.mapstruct.ap.test.ignore.AnimalMapper;
+import org.mapstruct.ap.test.ignore.Animal;
+import org.mapstruct.ap.test.ignore.AnimalDto;
+
+/**
+ * Test for ignoring properties during the mapping.
+ *
+ * @author Gunnar Morling
+ */
+public class AnimalTest {
```
Comment (gunnarmorling) [inline integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java]: Where is that one coming from?

File: integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    processorTypes = {
+        ProcessorType.ORACLE_JAVA_6,
+        ProcessorType.ORACLE_JAVA_7,
+        ProcessorType.ORACLE_JAVA_8,
+        ProcessorType.ORACLE_JAVA_9,
+        ProcessorType.ECLIPSE_JDT_JAVA_6,
+        ProcessorType.ECLIPSE_JDT_JAVA_7,
+        ProcessorType.ECLIPSE_JDT_JAVA_8
+})
+public class FullFeatureCompilationTest {
```
Comment (gunnarmorling) [inline integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationTest.java]: So what is the motivation for this, just compile all the test mappers on all the compilers and see whether there is no compilation error?

File: integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java [hunk 3]
```diff
@@ -43,11 +40,16 @@
 import org.junit.runner.notification.StoppedByUserException;
 import org.junit.runners.ParentRunner;
 import org.junit.runners.model.InitializationError;
+import org.mapstruct.itest.testutil.runner.ProcessorSuite.CommandLineEnhancer;
 import org.mapstruct.itest.testutil.runner.ProcessorSuite.ProcessorType;
 import org.mapstruct.itest.testutil.runner.ProcessorSuiteRunner.ProcessorTestCase;
 import org.mapstruct.itest.testutil.runner.xml.Toolchains;
 import org.mapstruct.itest.testutil.runner.xml.Toolchains.ProviderDescription;
 
+import static org.apache.maven.it.util.ResourceExtractor.extractResourceToDestination;
```
Comment (gunnarmorling) [inline integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java]: Could we try and keep the static imports first? It's the default in Eclipse and I'd prefer to keep it that way to avoid unneccessary line changes.

**Comment by gunnarmorling** [comment]
Ok, I guess it's fine :) Only that it adds ~20sec execution time to the build :/

File: integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationTest.java (hunk 2 above)
Comment (agudian) [inline integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationTest.java]: Exactly. As our processor tests check all kinds of crazy edge cases, we can have a higher coverage with the different compiler versions.
What's not covered is the check that each mapper is actually created correctly, i.e. it's doing the right thing. Perhaps we can come up with something to get closer to that as well.

File: integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java (hunk 3 above)
Comment (agudian) [inline integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java]: Yeah, sure. It was that kind of change that I got quite often in several files, so I thought you guys have them below the other imports as well - and thus I changed the setting locally. Just changed it back... Should I add a commit to organize all imports again?

File: integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java (hunk 1 above)
Comment (agudian) [inline integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java]: You don't recognize it? :grin: 
I copied it from the processor tests, to have at least one check on a generated mapper.

File: integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java (hunk 3 above)
Comment (gunnarmorling) [inline integrationtest/src/test/java/org/mapstruct/itest/testutil/runner/ProcessorSuiteRunner.java]: Yes, if you could do that globally that'd be great.

It may have been my fault actually, someone in Hibernate had started this
pattern so I adapted and maybe I used the wrong IDE style in MapStruct, too.

**Comment by agudian** [comment]
> Only that it adds ~20sec execution time to the build :/

Lucky you. It's four times that long on my machine, but I think it's worth it... :/

File: integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java (hunk 1 above)
Comment (gunnarmorling) [inline integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java]: But we have some like that in the other integration tests? Is there any advantage to adding this one?

**Comment by gunnarmorling** [comment]
> Lucky you. It's four times that long on my machine...

Doh...

Maybe we should add a switch for disabling that? But then one can just build the processor module for a quick build anyways... Overall, the execution time of the integrationtest module starts to hurt.

File: integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java (hunk 1 above)
Comment (agudian) [inline integrationtest/src/test/resources/fullFeatureTest/src/test/java/org/mapstruct/itest/simple/AnimalTest.java]: Hmm, yea... initially, I wanted to make sure the test actually generated any mappers - hitting the right relative directory was tricky :wink:

If it's alright with you, I'd like to keep it, to make sure the test continues to do what it's supposed to do. The test doesn't add any noticable time anyway.

**Comment by agudian** [comment]
Rebased and added the organize-imports.
